### PR TITLE
Fix logging tests that depends on each other

### LINKED ISSUE #1286: 0.16.0: pytetst is failing in random units when `pytest-randomly` is used
### Checklist

- [X] The bug is reproducible against the latest release or `master`.
- [X] There are no similar issues or pull requests to fix it yet.

### Describe the bug

Looks like pytest test suite depends on exact order of exact units. Consquence of that may be that on testing indyvidual unit by `pytest <file.py>::<unit>` it will produce false result.
Examples of the failing units:
(please ignore ipv6 unit failing because it is reported in another ticket)
```conosle
+ PYTHONPATH=/home/tkloczko/rpmbuild/BUILDROOT/python-uvicorn-0.16.0-2.fc35.x86_64/usr/lib64/python3.8/site-packages:/home/tkloczko/rpmbuild/BUILDROOT/python-uvicorn-0.16.0-2.fc35.x86_64/usr/lib/python3.8/site-packages
+ /usr/bin/pytest -ra
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.8.12, pytest-6.2.5, py-1.11.0, pluggy-0.13.1
benchmark: 3.4.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
Using --randomly-seed=3526222027
rootdir: /home/tkloczko/rpmbuild/BUILD/uvicorn-0.16.0, configfile: setup.cfg
plugins: shutil-1.7.0, virtualenv-1.7.0, mock-3.6.1, cov-2.12.1, anyio-3.3.4, forked-1.3.0, xdist-2.3.0, flaky-3.7.0, tornasync-0.6.0.post2, console-scripts-1.2.0, trio-0.7.0, timeout-2.0.1, asyncio-0.16.0, freezegun-0.4.2, flake8-1.0.7, pyfakefs-4.5.3, hypothesis-6.29.3, benchmark-3.4.1, profiling-1.7.0, datadir-1.3.1, regressions-2.2.0, randomly-3.8.0, rerunfailures-9.1.1, yagot-0.5.0, requests-mock-1.9.3, subtests-0.5.0, easy-server-0.8.0, django-4.5.2
collected 272 items

tests/middleware/test_proxy_headers.py ...........                                                                                                                   [  4%]
tests/test_cli.py ...........                                                                                                                                        [  8%]
tests/test_ssl.py ....                                                                                                                                               [  9%]
tests/middleware/test_message_logger.py ..                                                                                                                           [ 10%]
tests/importer/test_importer.py ......                                                                                                                               [ 12%]
tests/test_lifespan.py ...............                                                                                                                               [ 18%]
tests/test_auto_detection.py ...                                                                                                                                     [ 19%]
tests/middleware/test_logging.py ..........F                                                                                                                         [ 23%]
tests/test_default_headers.py ......                                                                                                                                 [ 25%]
tests/test_main.py .F......                                                                                                                                          [ 28%]
tests/protocols/test_http.py ............................................                                                                                            [ 44%]
tests/supervisors/test_multiprocess.py .                                                                                                                             [ 44%]
tests/middleware/test_wsgi.py .....                                                                                                                                  [ 46%]
tests/protocols/test_utils.py ......                                                                                                                                 [ 48%]
tests/test_config.py ................F.........F....................                                                                                                 [ 66%]
tests/protocols/test_websocket.py ................................................................                                                                   [ 89%]
tests/middleware/test_debug.py ....                                                                                                                                  [ 91%]
tests/supervisors/test_reload.py ........................                                                                                                            [100%]

[..]

========================================================================= short test summary info ==========================================================================
FAILED tests/middleware/test_logging.py::test_trace_logging_on_http_protocol[httptools] - AttributeError: 'Server' object has no attribute 'servers'
FAILED tests/test_main.py::test_run[ipv6] - httpx.ConnectError: All connection attempts failed
FAILED tests/test_config.py::test_log_config_default[use_colors_invalid_value] - assert True == None
FAILED tests/test_config.py::test_log_config_default[use_colors_not_provided] - assert True == None
====================================================================== 4 failed, 268 passed in 35.28s ======================================================================
```
```console
+ PYTHONPATH=/home/tkloczko/rpmbuild/BUILDROOT/python-uvicorn-0.16.0-2.fc35.x86_64/usr/lib64/python3.8/site-packages:/home/tkloczko/rpmbuild/BUILDROOT/python-uvicorn-0.16.0-2.fc35.x86_64/usr/lib/python3.8/site-packages
+ /usr/bin/pytest -ra
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.8.12, pytest-6.2.5, py-1.11.0, pluggy-0.13.1
benchmark: 3.4.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
Using --randomly-seed=459250520
rootdir: /home/tkloczko/rpmbuild/BUILD/uvicorn-0.16.0, configfile: setup.cfg
plugins: shutil-1.7.0, virtualenv-1.7.0, mock-3.6.1, cov-2.12.1, anyio-3.3.4, forked-1.3.0, xdist-2.3.0, flaky-3.7.0, tornasync-0.6.0.post2, console-scripts-1.2.0, trio-0.7.0, timeout-2.0.1, asyncio-0.16.0, freezegun-0.4.2, flake8-1.0.7, pyfakefs-4.5.3, hypothesis-6.29.3, benchmark-3.4.1, profiling-1.7.0, datadir-1.3.1, regressions-2.2.0, randomly-3.8.0, rerunfailures-9.1.1, yagot-0.5.0, requests-mock-1.9.3, subtests-0.5.0, easy-server-0.8.0, django-4.5.2
collected 272 items

tests/middleware/test_message_logger.py ..                                                                                                                           [  0%]
tests/protocols/test_websocket.py ................................................................                                                                   [ 24%]
tests/test_ssl.py ....                                                                                                                                               [ 25%]
tests/test_main.py .......F                                                                                                                                          [ 28%]
tests/supervisors/test_multiprocess.py .                                                                                                                             [ 29%]
tests/middleware/test_proxy_headers.py ...........                                                                                                                   [ 33%]
tests/test_lifespan.py ...............                                                                                                                               [ 38%]
tests/test_default_headers.py ......                                                                                                                                 [ 40%]
tests/test_auto_detection.py ...                                                                                                                                     [ 41%]
tests/middleware/test_debug.py ....                                                                                                                                  [ 43%]
tests/test_config.py .................F..........................F..                                                                                                 [ 60%]
tests/middleware/test_wsgi.py .....                                                                                                                                  [ 62%]
tests/supervisors/test_reload.py ........................                                                                                                            [ 71%]
tests/middleware/test_logging.py ....FF.FF.F                                                                                                                         [ 75%]
tests/protocols/test_http.py ............................................                                                                                            [ 91%]
tests/test_cli.py ...........                                                                                                                                        [ 95%]
tests/importer/test_importer.py ......                                                                                                                               [ 97%]
tests/protocols/test_utils.py ......                                                                                                                                 [100%]

[..]

 ========================================================================= short test summary info ==========================================================================
FAILED tests/test_main.py::test_run[ipv6] - httpx.ConnectError: All connection attempts failed
FAILED tests/test_config.py::test_log_config_default[use_colors_invalid_value] - assert True == None
FAILED tests/test_config.py::test_log_config_default[use_colors_not_provided] - assert False == None
FAILED tests/middleware/test_logging.py::test_trace_logging - IndexError: pop from empty list
FAILED tests/middleware/test_logging.py::test_trace_logging_on_ws_protocol[wsproto] - assert False
FAILED tests/middleware/test_logging.py::test_trace_logging_on_ws_protocol[websockets] - assert False
FAILED tests/middleware/test_logging.py::test_trace_logging_on_http_protocol[httptools] - AttributeError: 'Server' object has no attribute 'servers'
FAILED tests/middleware/test_logging.py::test_trace_logging_on_http_protocol[h11] - assert False
====================================================================== 8 failed, 264 passed in 34.28s ======================================================================
```
To diagnose that issue https://github.com/mrbean-bremen/pytest-find-dependencies/ needs to be used.
Teporaryy workaround: use `pytest -p no:randomly`

### Steps to reproduce the bug

1. Install `pytest-randomly`
2. run pytetst

### Expected behavior

None of the test sute units should be execution order dependent.

### Actual behavior

Fails in random units.

### Debugging material

_No response_

### Environment

N/A

### Additional context

None.

### PR DESCRIPTION
**Pull request by euri10**
Fixes #1286 or part of it, since I'm not sure yet if there are not other cases where parametrized tests are messing up with our test suite.

In this case the order of parameterized tests was important because the global `LOGGING_CONFIG` is not reset for each test : so we force it in the config using a fixture that is a deepcopy of the dict, so we use the original dict ofr each test and not the result of the previous one.

### CODE REVIEW

**Comment by euri10** [comment]
ok this seems to pass like that but still not with pytest-randomly, will look further later

**Comment by euri10** [comment]
ok can trigger the dependency bug on a rather small subset of tests
```
❯ pytest tests/middleware/test_logging.py  tests/supervisors/test_reload.py
========================================================================================================= test session starts =========================================================================================================
platform linux -- Python 3.9.9, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
Using --randomly-seed=1646219851
rootdir: /home/lotso/PycharmProjects/uvicorn, configfile: setup.cfg
plugins: asyncio-0.15.1, anyio-3.5.0, mock-3.6.1, randomly-3.11.0
collected 35 items                                                                                                                                                                                                                    

tests/supervisors/test_reload.py ........................                                                                                                                                                                       [ 68%]
tests/middleware/test_logging.py FF....FF.F.                                                                                                                                                                                    [100%]

============================================================================================================== FAILURES ===============================================================================================================
```

investigating why the reload tests seems to mess up the logging ones down the road...

**Comment by euri10** [comment]
smaller subset with
```
❯ pytest tests/middleware/test_logging.py  tests/supervisors/test_reload.py::TestBaseReload::test_should_detect_new_exclude_dirs
========================================================================================================= test session starts =========================================================================================================
platform linux -- Python 3.9.9, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
Using --randomly-seed=881466683
rootdir: /home/lotso/PycharmProjects/uvicorn, configfile: setup.cfg
plugins: asyncio-0.15.1, anyio-3.5.0, mock-3.6.1, randomly-3.11.0
collected 12 items                                                                                                                                                                                                                    

tests/supervisors/test_reload.py .                                                                                                                                                                                              [  8%]
tests/middleware/test_logging.py F..F.F..FF.                                                                                                                                                                                    [100%]

============================================================================================================== FAILURES ===============================================================================================================
```

File: tests/conftest.py [hunk 1]
```diff
@@ -128,3 +131,8 @@ def reload_directory_structure(tmp_path_factory: pytest.TempPathFactory):
     ext_file.touch()
 
     yield root
+
+
+@pytest.fixture(scope="function")
```
Comment (euri10) [inline tests/conftest.py:136]: using the log_config fixture on a function scope to ensure we get a clean one each test : that is useful when we mess up with caplog pytest internals in the caplog_for_logger context manager

File: requirements.txt [hunk 2]
```diff
@@ -23,6 +23,7 @@ coverage==5.5
 coverage-conditional-plugin==0.4.0
 httpx==1.0.0b0
 pytest-asyncio==0.15.1
+pytest-randomly==3.11.0
```
Comment (euri10) [inline requirements.txt]: not sure we should keep it, but that revealed subtle test suite bugs

**Comment by euri10** [comment]
ready for review !

**Comment by euri10** [comment]
if by chance @kloczek you get some time to test that branch that would be fantastic, I ran pytest suite for like 20 times randmly without triggering the bug, on the whole suite and on the smallest subset i could find but you never know

File: requirements.txt (hunk 2 above)
Comment (lovelydinosaur) [reply requirements.txt]: Hmm... very, very neat bit of tooling, tho I'd probably suggest that we *don't* have it in for our regular CI.

**Review by lovelydinosaur — COMMENTED** [review]
Fantastic! Let's drop `pytest-randomly` - but otherwise great work.

File: tests/test_config.py [hunk 3]
```diff
@@ -32,11 +31,6 @@ def mocked_logging_config_module(mocker: MockerFixture) -> MagicMock:
     return mocker.patch("logging.config")
 
 
-@pytest.fixture(scope="function")
-def logging_config() -> dict:
-    return deepcopy(LOGGING_CONFIG)
-
-
```
Comment (euri10) [inline tests/test_config.py:39]: I put it in conftest but can put it back here to minimize code change if needed

**Review by lovelydinosaur — APPROVED** [review]

**Comment by kloczek** [comment]
> if by chance @kloczek you get some time to test that branch that would be fantastic, I ran pytest suite for like 20 times randmly without triggering the bug, on the whole suite and on the smallest subset i could find but you never know

Just manage to dig out to bottom of my inbox and found email about that comment 😄 
In mean time 0.17.6 has been released.
Just tested that version updating my rpm spec file with `uvicorn` module.
- pytest fails with latest `pytest-asyncio` module which is needed
```console
+ PYTHONPATH=/home/tkloczko/rpmbuild/BUILDROOT/python-uvicorn-0.17.6-2.fc35.x86_64/usr/lib64/python3.8/site-packages:/home/tkloczko/rpmbuild/BUILDROOT/python-uvicorn-0.17.6-2.fc35.x86_64/usr/lib/python3.8/site-packages
+ /usr/bin/pytest -ra -p no:randomly
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/_pytest/main.py", line 264, in wrap_session
INTERNALERROR>     config._do_configure()
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/_pytest/config/__init__.py", line 995, in _do_configure
INTERNALERROR>     self.hook.pytest_configure.call_historic(kwargs=dict(config=self))
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_hooks.py", line 277, in call_historic
INTERNALERROR>     res = self._hookexec(self.name, self.get_hookimpls(), kwargs, False)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_manager.py", line 80, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_callers.py", line 60, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_result.py", line 60, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pluggy/_callers.py", line 39, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/pytest_asyncio/plugin.py", line 191, in pytest_configure
INTERNALERROR>     config.issue_config_time_warning(LEGACY_MODE, stacklevel=2)
INTERNALERROR>   File "/usr/lib/python3.8/site-packages/_pytest/config/__init__.py", line 1342, in issue_config_time_warning
INTERNALERROR>     warnings.warn(warning, stacklevel=stacklevel)
INTERNALERROR> DeprecationWarning: The 'asyncio_mode' default value will change to 'strict' in future, please explicitly use 'asyncio_mode=strict' or 'asyncio_mode=auto' in pytest configuration file.
```
This can be fixed by patch like below:
```patch
--- a/pytest.ini
+++ b/pytest.ini
@@ -0,0 +1,2 @@
+[pytest]
+asyncio_mode = auto
```

**Comment by euri10** [comment]
we are still using a lower version @kloczek 
https://github.com/encode/uvicorn/blob/b8e65e49bfadf6d5ea30e56d0cd9f47329de606c/requirements.txt#L25
this will indeed fix the issue on newer pytest-asyncio version but opens for us a set of issue on windows we didn't fix yet

**Comment by kloczek** [comment]
> this will indeed fix the issue on newer pytest-asyncio version but opens for us a set of issue on windows we didn't fix yet

Pls correct me if I'm wrong. I tink that this pytest.ini entry probably can be added even now and should not affect use still lower version of the `pytest-asyncio` 🤔

**Pull request merged**